### PR TITLE
GH-654 As a developer I want N4JSC to install dependencies based on manifest file.

### LINKED ISSUE #654: As a developer i want N4JSC to install dependencies based on manifest file.
N4JSC.jar should provide functionality similar to https://github.com/eclipse/n4js/issues/512, i.e. headless compiler should fetch dependencies needed based on the manifest files. This would reduce burden of maintaining same dependency information in both manifest file and targetplatform file.

### PR DESCRIPTION
**Pull request by yoosiba**
resolves #654

### CODE REVIEW

File: tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/N4jscBase.java [hunk 1]
```diff
@@ -492,6 +503,22 @@ public SuccessExitStatus doMain(String... args) throws ExitCodeException {
 				// clean without compiling anything.
 				clean();
 			} else {
+				if (installMissingDependencies) {
+					Map<String, String> dependencies = dependencyHelper.discoverMissingDependencies(projectLocations,
+							srcFiles);
+					if(verbose){
```
Comment (mmews-n4) [inline tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/N4jscBase.java]: is autoformat turned on?

File: tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/DependenciesHelper.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+	// TODO GH-651 copied org.eclipse.n4js.generator.headless.HeadlessHelper.collectAllProjectPaths(List<File>)
+	static ArrayList<File> collectAllProjectPaths(List<File> absProjectRoots) {
+		ArrayList<File> pDir = new ArrayList<>();
+		for (File projectRoot : absProjectRoots) {
+			Arrays.asList(projectRoot.listFiles(f -> {
+				return f.isDirectory(); // all directrories
+			}))//
+					.stream() //
+					.filter(f -> {
+						File[] list = f.listFiles(f2 -> f2.getName().equals(IN4JSProject.N4MF_MANIFEST));
```
Comment (mmews-n4) [inline tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/DependenciesHelper.java]: my usual comment about understandability here

Comment (yoosiba) [reply tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/DependenciesHelper.java]: yes, I have refactored it in the GH-651

File: tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/N4jscBase.java (hunk 1 above)
Comment (yoosiba) [reply tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/N4jscBase.java]: In a words of Severus Snape: "Always"

**Comment by yoosiba** [comment]
`InstallFromManifestCompileRunN4jscExternalImportsTest`

**Pull request merged**